### PR TITLE
[flang][openacc] Lower parallel and data constructs

### PR DESCRIPTION
**Pull request by clementval**
This PR enable the lowering of the parallel and data construct and update the loop construct lowering. 

This cannot be merged before https://reviews.llvm.org/D87991 has been merged upstream and the `fir-dev` branch has been rebased.

### CODE REVIEW

**Comment by clementval** [comment]
https://reviews.llvm.org/D87991 has landed. So this PR just need to wait on the rebase of fir-dev before being able to be merged.

File: flang/lib/Lower/OpenACC.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/OpenACC.cpp:34]: Do we expect that we'll almost never have more than 2 operands?

We should consider using SmallVectorImpl as the interface though. Like here: https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/IR/FunctionImplementation.h#L57-L62

File: flang/lib/Lower/OpenACC.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/OpenACC.cpp:268]: This seems to start a really big block of code. Do you want to split it up into smaller chunks?

File: flang/test/Lower/OpenACC/acc-data.f90
Comment (schweitzpgi) [inline flang/test/Lower/OpenACC/acc-data.f90:8]: I think most of our tests leave the `%` out of the regexp as an anchor for the match.

**Review by schweitzpgi — APPROVED** [review]
LGTM.

I'll be doing another rebase soon.

File: flang/test/Lower/OpenACC/acc-data.f90
Comment (clementval) [reply flang/test/Lower/OpenACC/acc-data.f90:8]: Right. I'll update that.

File: flang/lib/Lower/OpenACC.cpp
Comment (clementval) [reply flang/lib/Lower/OpenACC.cpp:268]: Yes I plan to do that since this will likely to be reused when I'll work on lowering `!$acc parallel loop`.

File: flang/lib/Lower/OpenACC.cpp
Comment (clementval) [reply flang/lib/Lower/OpenACC.cpp:34]: Well I guess we will have cases where we have more for sure. I'm fine with updating it with the `SmallVectorImpl`.

File: flang/test/Lower/OpenACC/acc-data.f90
Comment (clementval) [reply flang/test/Lower/OpenACC/acc-data.f90:8]: I checked in the mlir directory and when the value is captured the `%` is kept in the regex. So Ieft it as well. 

https://github.com/llvm/llvm-project/blob/61ac58e10ae1fbb18416cac20119725803b9ad6f/mlir/test/Conversion/SCFToStandard/convert-to-cfg.mlir#L229

**Comment by clementval** [comment]
@schweitzpgi Should be good to be merged.

**Pull request merged**
